pkg/custompluginmonitor/custom_plugin_monitor.go

```diff
@@ -202,6 +252,19 @@ func (c *customPluginMonitor) generateStatus(result cpmtypes.Result) *types.Stat
 						condition.Reason,
 						timestamp,
 					))
+
+					// update metrics only when condition reason changes.
+					if condition.Reason != result.Rule.Reason && *c.config.EnableMetricsReporting {
+						err := problemmetrics.GlobalProblemMetricsManager.IncrementProblemCounter(result.Rule.Reason, 1)
+						if err != nil {
+							glog.Errorf("Failed to update problem counter metrics for %q: %v", result.Rule.Reason, err)
+						}
+						err = problemmetrics.GlobalProblemMetricsManager.SetProblemGauge(result.Rule.Condition, result.Rule.Reason, true)
+						if err != nil {
+							glog.Errorf("Failed to update problem gauge metrics for problem %q, reason %q: %v",
+								result.Rule.Condition, result.Rule.Reason, err)
+						}
+					}
 				}
 
```

Comment:
Generate counter and gauge at the end of the function based on the final events and condition.

The logic is too complex, the code makes it even harder to maintain.

Reply:
And let's do the same for system log monitor.

Reply:
Ah this is a very good idea! Thanks for noticing that1
Done.